### PR TITLE
Declare SILE.scratch.headers in twoside only

### PR DESCRIPTION
**Pull request by jodros**
#2086

### CODE REVIEW

File: classes/bible.lua [hunk 1]
```diff
@@ -6,10 +6,6 @@ local plain = require("classes.plain")
 local class = pl.class(plain)
 class._name = "bible"
```
Comment (Omikhleia) [inline classes/bible.lua:7]: This class is a joke.

When will we kill it?

File: packages/twoside/init.lua [hunk 2]
```diff
@@ -72,6 +72,9 @@ function package:_init (options)
    end)
    self.class:registerHook("newpage", spreadHook)
    self.class:registerHook("newpage", switchPage)
+   if not SILE.scratch.headers then
+      SILE.scratch.headers = {}
+   end
```
Comment (Omikhleia) [inline packages/twoside/init.lua:77]: The only place "twoside" uses headers scratch are in `\open-spread` -- but is it approached as it should, there?

Is even this command `\open-spread` in the right package/module? Is the solution general? (The answer is obviously "no" to me, so I'd like seeing good rationales for it...)

File: classes/bible.lua (hunk 1 above)
Comment (jodros) [reply classes/bible.lua:7]: > This class is a joke.

Why?

File: packages/twoside/init.lua (hunk 2 above)
Comment (jodros) [reply packages/twoside/init.lua:77]: > Is even this command \open-spread in the right package/module?

I have no idea... Where do you think it should be?

File: classes/bible.lua (hunk 1 above)
Comment (Omikhleia) [reply classes/bible.lua:7]: Er? What is supposed to be a "bible" class?

Food for thought: I did typeset many bibles (https://github.com/Freely-Given-org/BibleTypesetter/pull/3 -- Still pending any sort of review and approval, which is sad, but anyway) without needing a "bible" **class**.

So what is supposed to be a "bible" class in layman terms?

**Comment by RobH123** [comment]
On 23/07/24 09:24, Omikhleia wrote:
> Food for thought: I did typeset many bibles 
> (Freely-Given-org/BibleTypesetter#3 
> <https://github.com/Freely-Given-org/BibleTypesetter/pull/3> -- Still 
> pending any sort of review and approval, which is sad, but anyway) 
> without needing a "bible" *class*.
Yes @Omikhleia, I apologise for the slow reaction, but it's not 
forgotten and still on my to-do list. I'm still very interested in 
evaluating your work and hopefully extending/expanding it, it's just 
that other priorities still prevail for now.

File: packages/twoside/init.lua (hunk 2 above)
Comment (Omikhleia) [reply packages/twoside/init.lua:77]: > I have no idea... Where do you think it should be?

My deep feeling is that only the **class** knows how to handle page breaks and what do to with "empty" pages, (based on the reason why it it introducing them with so-called spreads).

It's hard to explain with the "limited" book class provided by SILE, but I can try, based on my experience designing books with my own resilient.book class (again, it does have some flaws I am aware of, however)[^1]:

 - Consider a `\part` (yep, resilient has these) followed by a chapter, and both open on odd pages, so a blank even page is gonna be inserted in between. Are you in a "novel" or in a "technical documentation"? Why I am asking? 'Cause in a novel, usually, the book name is on even pages (so it's not a problem to repeat it, at least), but in a technical documentation, the chapter name is usually on even pages (so it has to be cleared after a part...)
 - Consider a `\chapter` followed by another chapter, also opening on an odd page[^2], and a blank even page might be needed. Same question, but trickier: to which chapter belongs the empty page? Should it have a header (not to say a footer), and which one? There's no simple answer, except that "it depends" on some publisher preferences...

And here we're indeed discussing parts, chapters, etc. -- that is, things that are (currently) a class-design topic.

[^1]: If you read its doc, you'll see for instance that resilient uses its own header logic, and that it has provision for "novel" and "technical" headers... Though the underlying implementation is a bit of a hack.
[^2]: Though that ought to be configurable: Some books _do_ allow chapters to begin on even pages... No wonder it took me years to design my resilient.book solution, as all styling decisions are configurable -- and suddenly, all sort of edge cases pop up :)

Comment (jodros) [reply packages/twoside/init.lua:77]: Thanks for your explanation, I got your point.

Well, I did this PR without ever thinking about `twoside` not being the most suitable place for `open-spread` or that scratch is about to deprecate... For me it was way simpler.

File: classes/bible.lua (hunk 1 above)
Comment (alerque) [reply classes/bible.lua:7]: Bible typesetting comes with a standard set of features over and above what most books require. Defaulting to grid typesetting, having headers that keep track of the verse at start and end of pages for use in headers, multiple types of footnotes/cross references, etc. The *bible* class should build on the book class with a few default features available that cover common needs that aren't so common in regular books. Of course you don't have to use it, it's just a starting point.

Additionally it's where I'd put stuff to cover default processing functions for OSIS/USFM/whatever tags that don't have any relevance to non-Bible content.
